logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,173 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-            return null;
-        } else if (variable.isThing()) {
-            if (variable.asThing().isa().isPresent()) return variable.asThing().isa().get().type();
-            return null;
-        } else throw GraknException.of(ILLEGAL_STATE);
-    }
-
-    private void ensureResolvedConformToTheirSuper(Variable variable, Set<Variable> visited) {
-        if (variable == null || visited.contains(variable) || variable.reference().isLabel()) return;
-        visited.add(variable);
-        ensureResolvedConformToTheirSuper(above(variable), visited);
-        removeVarsViolatingSuper(variable);
-    }
-
-    private void removeVarsViolatingSuper(Variable subVar) {
-        TypeVariable supVar = above(subVar);
-        if (supVar == null) return;
-        if (supVar.reference().isLabel()) return;
-
-        Set<Label> supLabels = supVar.resolvedTypes();
-        Set<Label> subLabels = subVar.resolvedTypes();
-        if (supLabels.isEmpty()) return;
-
-        if (subLabels.isEmpty()) {
-            Set<Label> subResolvedOfSupLabels = supLabels.stream().flatMap(label -> getType(label).getSubtypes())
-                    .map(Type::getLabel).collect(Collectors.toSet());
-            subVar.addResolvedTypes(subResolvedOfSupLabels);
-            return;
-        }
+    public Conjunction resolveVariables(Conjunction conjunction) {
+        resolveLabels(conjunction);
+        TraversalConstructor traversalConstructor = new TraversalConstructor(conjunction, conceptMgr);
 
-        Set<Label> temp = new HashSet<>(subLabels);
-        for (Label label : temp) {
-            Type ResolvedType = getType(label);
-            while (ResolvedType != null && !supLabels.contains(ResolvedType.getLabel())) {
-                ResolvedType = ResolvedType.getSupertype();
-            }
-            if (ResolvedType == null) {
-                subVar.removeResolvedType(label);
-                if (subVar.resolvedTypes().isEmpty()) subVar.setSatisfiable(false);
-            }
-        }
-    }
+        Map<Reference, Set<Label>> resolvedLabels = runTraversalEngine(traversalConstructor);
+        if (resolvedLabels.isEmpty()) throw GraknException.of(SCHEMATICALLY_UNSATISFIABLE_CONJUNCTION, conjunction);
+        resolvedLabels.forEach((ref, labels) -> traversalConstructor.getVariable(ref).addResolvedTypes(labels));
+        findVariablesWithNoHints(conjunction);
 
-    private void addInferredSubLabels(TypeVariable variable, Set<Label> resolvedLabels) {
-        variable.addResolvedTypes(resolvedLabels);
+        return conjunction;
     }
 
-    private void addInferredIsaLabels(ThingVariable variable, Set<Label> resolveLabels) {
-        //TODO: use .getType(label) once ConceptManager can handle labels
-        resolveLabels.removeIf(label -> traversalEng.graph().schema().getType(label).isAbstract());
-        variable.addResolvedTypes(resolveLabels);
+    private void findVariablesWithNoHints(Conjunction conjunction) {
```

Comment:
I'm deliberately using "hints" to say that the resolved types give no additional information. So the method is shorthand for:

"find out which variables have Resolve Types which give no hint their possibilities".

However, since you've written this comment it must not have been clear, so I will think of something more appropriate.

Reply:
Changed to `clearResolvedTypesWhichProvideNoInfo`.